libkbfs/keybase_daemon_rpc.go

```diff
@@ -255,47 +277,28 @@ func (k *KeybaseDaemonRPC) AddProtocols(protocols []rpc.Protocol) {
 	} else {
 		k.protocols = protocols
 	}
+
+	// If we are already connected, register these protocols.
+	k.lock.RLock()
+	defer k.lock.RUnlock()
+	if k.server != nil {
+		for _, p := range protocols {
+			err := k.server.Register(p)
+			if err != nil {
+				k.log.Warning("add protocol %q error", p.Name)
+			}
+		}
+	}
 }
 
 // OnConnect implements the ConnectionHandler interface.
 func (k *KeybaseDaemonRPC) OnConnect(ctx context.Context,
 	conn *rpc.Connection, rawClient rpc.GenericClient,
 	server *rpc.Server) error {
+	k.lock.Lock()
```

Comment:
I'm slightly nervous about holding a lock while we do the Hello RPC below (in case of deadlocks, etc) but I think it's pretty safe in this case and probably for the best.  So maybe just add a comment explaining why that's safe?

Reply:
Good point. I ended up just holding a R lock at beginning and take the lock again at the end.